viewing_party/party.py

```diff
@@ -1,23 +1,151 @@
-    pass
 
+    if title == None or genre == None or rating == None:
+        return None
+    else:
+        movie = {
+                        "title": title,
+                        "genre": genre,
+                        "rating": rating,
+                        }
+        return movie
+    
+
+def add_to_watched(user_data, movie):
+   
+    user_data["watched"].append(movie)
+    return user_data
+
+
+def add_to_watchlist(user_data, movie):
+    
+    user_data["watchlist"].append(movie)
+    return user_data
+
+
+def watch_movie(user_data, title):
+
+    to_watch_list = user_data["watchlist"]
+    already_watched = user_data["watched"]
+
+    for i in range(len(to_watch_list)):
+        if to_watch_list[i]["title"] == title:
+            already_watched.append(to_watch_list[i])
+            to_watch_list.remove(to_watch_list[i])
+    return user_data
+    
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+   
+    already_watched = user_data["watched"]
+    ratings = 0
+    for i in range(len(already_watched)):
+        ratings += already_watched[i]["rating"]
+        avr_rating = (ratings/(len(already_watched)))
+    if already_watched == []:
+        avr_rating = 0
+        
+    return avr_rating
+
+def get_most_watched_genre(user_data):
 
+     already_watched = user_data["watched"]
+     all_genres = []
+
+     for i in range(len(already_watched)):
+        all_genres.append(already_watched[i]["genre"])
+        top_genre = [word for word, word_count in Counter(all_genres).most_common(1)]
+        top_genre = top_genre[0]
+        return top_genre
+     if already_watched == []:
+        top_genre = None
+     return top_genre
+     
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+def get_unique_watched(user_data):
+    
+    friends_not_watched = []
+    friends_watched = []
+    user_watched = user_data["watched"]
+
+    for movie in user_watched:
+        friends_not_watched.append(movie)
+
+    for friend_list in user_data["friends"]:
+        for friend in friend_list["watched"]:
+            friends_watched.append(friend)
+    for i in friends_watched:
+        if i in friends_not_watched:
+            friends_not_watched.remove(i)
+    return friends_not_watched
+
+def get_friends_unique_watched(user_data):
+   
+    only_friends_watched = []
+    already_watched = []
+    user_watched = user_data["watched"]
+
+    for friend_list in user_data["friends"]:
+        for friend in friend_list["watched"]:
+            if friend not in only_friends_watched:
+                only_friends_watched.append(friend)
+   
+    for movie in user_watched:
+        already_watched.append(movie)
+    
+    for i in already_watched:
+        if i in only_friends_watched:
+            only_friends_watched.remove(i)

```

Comment:
Good work in using a `for..in` loop. `i` is a very generic name, we should try to use descriptive variable names where we can! The most beautiful thing about python is its readability! Programming is like writing a story, all characters deserve relevant and meaningful names ✨ So in this case, we can rewrite this function like:

```python
for user_movie in already_watched:
    if user_movie in only_friends_watched:
         only_friends_watched.remove(user_movie)
```